network/cert_verifier.go

```diff
@@ -0,0 +1,124 @@
+package network
+
+import (
+	"crypto/x509"
+	"errors"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+)
+
+// CertVerifier is a generic certificate verifier
+type CertVerifier interface {
+	Verify(chain []*x509.Certificate) (int, error)
```

Comment:
why so difficult and return count of verifies?) we can return only errors and make several wrappers for such algorithms which will be related on confirm count
```
type CertVierifier interface {
  VerifyPeerCertificate func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
}

type OCSPVerifier struct{}
func (verifier *OCSPVerifier) VerifyPeerCertificate func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {return nil}

type CRLVerifier struct{}
func (verifier *CRLVerifier) VerifyPeerCertificate func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {return nil}

// we can use this implementation for at least and for verifyAll
type AtLeastXVerifier struct {
  verifiers []CertVierifier
  minCount int
}
func NewAtLeastXVerifier(x int, verifiers []Verifiers)(AtLeastXVerifier, error){return nil}
func (verifier *AtLeastXVerifier) VerifyPeerCertificate func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
  count := 0
  for _, v := range verifier.verifiers {
    if err := v.VerifyPeerCertificate(rawCerts, verifiedChains); err != nil {
      count++
    }
  }
  if count < verifier.minCount {
    return ErrNotEnoughVerificationCount
  }
}

func NewVerifyAllVerification(verifiers []CertVerifier)(*AtLeastXVerifier, error){
  // reuse AtLeastXVerifier as special case
  return NewAtLeastXVerifier(len(verifiers), verifiers)
}
```

After that we can configure to use only OCSPVerifier, only CRLVerifier, or combine both with special requirements like verify all or X verifiers. And move logic of counting to separate implementations which implement same interface

Reply:
Okay, I'd support not returning the successful verification count.

While it seems like a good thing to write into logs, I don't think Acra can make any decisions based on this number. When the verifier has been configured and is used, you don't even know how many internal verifiers there are inside. So you can't really tell what verification count is normal and what is not.

The only place where you can – sorta-kinda – look and act on the verification behavior is *the implementation* of the verifier: composite verifiers know how many direct children they have and which of them fail. However, they have no clue how many child verifiers their children have, just like the caller of the composite verifier does not know it.

So I concur with @Lagovas here and support returning an error alone. This makes verification a clear-cut process:

- Either the certificate is okay and we allow the connection.
- Or the certificate is now okay and we deny the connection.

There is no third option of "well, maybe" when some sub-verifiers don't return a conclusive result and the composite verifiers defers the decision to the caller by returning `nil` error but not-quite-successful verification count (which the caller cannot distinguish).

If only an error is returned, the verifier is responsible for making the decision whether it's okay or not to trust the certificate. The conditions of trust are configurable by the administrator.